infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -56,6 +53,7 @@
  */
 @Singleton
 public class KubernetesNamespaceFactory {
```

Comment:
wow the logic in this class is HUGE!  imho it's time to think how to split it.. maybe using strategy pattern (legasy vs new rules etc)

Reply:
yes, I lold at this `boolean newLogic =`

Reply:
Yes, this whole class is a mess. The naming was much more confusing before I tried to make it at least a little bit more approachable - `create()` would only instantiate an object that is able to access a namespace as just one example. We used the `EnvironmentContext` to get the current user on codepaths that don't originate in the REST API, etc.

I agree we should think about splitting this class. Just not sure we should do it at the same time as changing the logic.

I didn't find `newLogic` too funny when I had to write it to handle back compat ;) Strategy pattern would be good, but that would probably require quite a lot of change. The split between old and new logic is based on existence of a namespace - we cannot detect that in the constructor. Also the split is needed in a couple of methods that are hooked to a much larger workflow way beyond this class.

I guess we have different thresholds for code ugliness, but I still find it better to try and make more sense of this class as it is in the broader picture than try to rearchitect the workspace provisioning.

Reply:
https://github.com/eclipse/che/issues/15250